vsphere/resource_vsphere_host.go

```diff
@@ -0,0 +1,683 @@
+		_, new := d.GetChange("connected")
+		desiredConnectionState = new.(bool)
+	} else {
+		desiredConnectionState = d.Get("connected").(bool)
+	}
+
+	hostID := d.Id()
+	hostObject, err := hostsystem.FromID(client, hostID)
+	if err != nil {
+		return err
+	}
+
+	actualConnectionState, err := hostsystem.GetConnectionState(hostObject)
+	if err != nil {
+		return err
+	}
+
+	// Have there been any changes that warrant a reconnect?
+	reconnect := false
+	connectionKeys := []string{"hostname", "username", "password", "thumbprint"}
+	for _, k := range connectionKeys {
+		if d.HasChange(k) {
+			reconnect = true
+			break
+		}
+	}
+
+	// Decide if we're going to reconnect or not
+	reconnectNeeded, err := shouldReconnect(d, meta, actualConnectionState, desiredConnectionState, reconnect)
+	if err != nil {
+		return err
+	}
+
+	switch reconnectNeeded {
+	case 1:
+		err := handleReconnect(d, meta)
+		if err != nil {
+			return err
+		}
+	case -1:
+		err := handleDisconnect(d, meta)
+		if err != nil {
+			return err
+		}
+	case 0:
+		break
+	}
+
+	mutableKeys := map[string]func(*schema.ResourceData, interface{}, interface{}, interface{}) error{
+		"license":     modifyLicense,
+		"cluster":     modifyCluster,
+		"maintenance": modifyMaintenanceMode,
+		"lockdown":    modifyLockdownMode,
+	}
+	for k, v := range mutableKeys {
+		log.Printf("[DEBUG] Checking if key %s changed", k)
+		if !d.HasChange(k) {
+			continue
+		}
+		log.Printf("[DEBUG] Key %s has change, processing", k)
+		old, new := d.GetChange(k)
+		err := v(d, meta, old, new)
+		if err != nil {
+			return fmt.Errorf("error while updating %s: %s", k, err)
+		}
+	}
+	return resourceVsphereHostRead(d, meta)
+}
+
+func resourceVsphereHostDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*VSphereClient).vimClient
+	hostID := d.Id()
+
+	hs, err := hostsystem.FromID(client, hostID)
+	if err != nil {
+		return err
+	}
+
+	connectionState, err := hostsystem.GetConnectionState(hs)
+	if err != nil {
+		return err
+	}
+
+	if connectionState != types.HostSystemConnectionStateDisconnected {
+		// We cannot put a disconnected server in maintenance mode.
+		err = hostsystem.EnterMaintenanceMode(hs, int(defaultAPITimeout), true)
+		if err != nil {
+			return fmt.Errorf("error while putting host to maintenance mode: %s", err.Error())
+		}
+	}
+
+	task, err := hs.Destroy(context.TODO())
+	if err != nil {
+		return err
+	}
+	err = task.Wait(context.TODO())
+	if err != nil {
+		return fmt.Errorf("Error while waiting for host (%s) to be removed: %s", hostID, err)
+	}
+
+	var to mo.Task
+	err = task.Properties(context.TODO(), task.Reference(), nil, &to)
+	if err != nil {
+		log.Printf("[DEBUG] Failed while getting task results: %s", err)
+		return err
+	}
+
+	if to.Info.State != "success" {
+		return fmt.Errorf("Error while removing host(%s): %s", hostID, to.Info.Error)
+	}
+	return nil
+}
+
+func modifyLockdownMode(d *schema.ResourceData, meta, old, new interface{}) error {
```

Comment:
This is more a matter of just keeping a consistent style, but I'd recommend something along the lines of `resourceVSphereHostUpdateLockdownMode`. Its kind of long, but extra descriptive.

Reply:
👍